### PR TITLE
8303972: (zipfs) Make test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java independent of the zip command line

### PR DESCRIPTION
**Pull request by eirbjo**
Please review this PR which brings the currenly problem-listed test TestLocOffsetFromZip64EF back to life.

Instead of calling out to the `zip` command line, we produce a small-sized Zip64 file in the test itself. This file has the features  required to reproduce the ZipFileSystem issue, namely that the 'INFO-ZIP extended timestamp' field must come before  the 'Zip64 extended information'  field. (This would casue ZipFileSystem to read the LOC at position 0xFFFFFFFF).

This speed up the test (from 50s to 3s on my Macbook Pro), saves 4 GB disk space during builds and removes a brittle dependency on the `zip` command in OS/distros.

Seee [JDK-8301183](https://bugs.openjdk.org/browse/JDK-8301183) for details on the problem-listing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8303972](https://bugs.openjdk.org/browse/JDK-8303972): (zipfs) Make test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java independent of the zip command line (**Enhancement** - P4)
 * [JDK-8301183](https://bugs.openjdk.org/browse/JDK-8301183): (zipfs) jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java failing with ZipException:R0 on OL9 (**Bug** - P4)

### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12979/head:pull/12979` \
`$ git checkout pull/12979`

Update a local copy of the PR: \
`$ git checkout pull/12979` \
`$ git pull https://git.openjdk.org/jdk.git pull/12979/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12979`

View PR using the GUI difftool: \
`$ git pr show -t 12979`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12979.diff">https://git.openjdk.org/jdk/pull/12979.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12979#issuecomment-1463982477)

### CODE REVIEW

**Comment by eirbjo** [comment]
Converting to draft while I investigate why the updated test does not actually catch the regression

File: test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java
Comment (AlanBateman) [inline test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java:136]: Can you use FileChannel.open and specify SPARSE in the set of open options, I think that would make it clearer that the file is sparse and remove the need for SparseOutputStream.

Comment (eirbjo) [reply test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java:136]: Not sure I understand. I could use FileChannel.open with SPARSE (which btw is ignored by the implementation:), but that doesn't give me an OutputStream I can pass to ZipOutputStream?

I could use Channels.newOutputStream, but that would not create holes in the sparse files?

My current thinking is that we need the SparseOutputStream to detect that ZIpOutputStream is writing empty bytes and replace that with `channel.position(channel.position() + len)`

**Review by LanceAndersen — COMMENTED** [review]
This test was leveraging a file created by info-zip intentionally so care needs to be taken to ensure that the zip that is being created is similar LOC and CEN output

File: test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java
Comment (AlanBateman) [reply test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java:136]: > Not sure I understand. I could use FileChannel.open with SPARSE (which btw is ignored by the implementation:)

That is true on platforms where files can be sparse by default. On Windows you'll see this translates to create the file as sparse, it needs to be opt-in when creating the file.

**Comment by eirbjo** [comment]
> This test was leveraging a file created by info-zip intentionally so care needs to be taken to ensure that the zip that is being created is similar LOC and CEN output

Yes, I just noticed that and currently looking into it.  I think in particular it is sensitive to the ordering of the extended time stamp and the Zip64 entries, which is not the same as ZipOutputStream.

**Comment by eirbjo** [comment]
While it is possible to use a sparse file here, there is a problem that when ZipFileSystem reads inside a 'hole', it will skip ahead to the next data which turned out to be a LOC header. We could  carefully select an entry size such that 0XFFFFFFFF does not not fall into in a hole, but that feels a bit too magic.

Instead, I opted for creating a small ZIP file with a Zip64-formatted CEN. Since this issue is sensitive to ordering of extra fields, the 'extended timestamp' needs to come before the 'Zip64 extended information field'.

The way I solve this is to add an opaque extra field, right sized for 'Zip64'. This uses a tag ID unknown to ZipEntry.setExtra, such that it will not be parsed as Zip64 but simply passed through. ZipOutputStream will output this after the extended timestamp.

After the ZIP is produced, the Zip64 entry is updated with the correct tag.

I have verified that this causes ZipFileSystem (before JDK-8255380) to attempt to read the LOC from 0xFFFFFFFF. With a small file size, this causes a slightly different error message because of the attempt to read beyond the end of the file.

```
----------System.out:(160/10450)----------
config TestLocOffsetFromZip64EF.setUp(): success
test TestLocOffsetFromZip64EF.walkZipFSTest({}): failure
java.util.zip.ZipException: loc: reading failed
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem$Entry.readExtra(ZipFileSystem.java:3127)
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem$Entry.readCEN(ZipFileSystem.java:2831)
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem$Entry.<init>(ZipFileSystem.java:2767)
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.getFileAttributes(ZipFileSystem.java:550)
```  

While the symptom is slightly different, the cause is the same: ZipFileSystem tries to read the LOC at offset 0xFFFFFFFF because it has not read the Zip64 entry yet.

File: test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java
Comment (eirbjo) [reply test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java]: I opted to create a small-sized Zip64 file instead (see below). This file does not need to be sparse.

Comment (eirbjo) [reply test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java]: > That is true on platforms where files can be sparse by default. On Windows you'll see this translates to create the file as sparse, it needs to be opt-in when creating the file.

Thanks for this info, Alan!

While this PR no longer makes use of sparse files, I have applied your comments about FileChannel.open for NTFS in another PR (which does make use of sparse files):

https://github.com/openjdk/jdk/pull/12948

File: test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java
Comment (LanceAndersen) [inline test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java:218]: I would probably make 0x1 a constant like the unknown tag.

It would be good to beef up the comment as the byte array written by ZipEntry::setExtra is set to the unknown tag for the Tag and the length is set to 28 which includes all of the Zip64 extra fields

File: test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java
Comment (LanceAndersen) [inline test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java:173]: The length of the zip64 extra field is going to be 28.   I would suggest making this a bit clearer and adding a field/constant which is set to 28 for clarity to make it easier for those who might look at this code later

**Review by LanceAndersen — COMMENTED** [review]
Thank you for your work here Eirik, I think we are close.

I think I would include an additional entry in the zip just as an extra sanity check that we can navigate an additional entry

It is also worth noting that the LOC will still contain an extra field with the unknown tag of 0x9902 while the CEN is being changed to use Zip64, 0x1

A few more minor comments below.

I ran the current version of the test and it runs quickly on all platforms via mach5

Best
Lance

**Comment by eirbjo** [comment]
> I think I would include an additional entry in the zip just as an extra sanity check that we can navigate an additional entry

Added an additional entry as suggested and checked that is was navigated.

> It is also worth noting that the LOC will still contain an extra field with the unknown tag of 0x9902 while the CEN is being changed to use Zip64, 0x1

This stray opaque extra fild should not be any issue, but just in case, I moved the ZipFIle.setExtra until after the LOC is written, so now only the CEN will have the Zip64 extra field.

File: test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java
Comment (eirbjo) [reply test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java]: > I would probably make 0x1 a constant like the unknown tag.

Extracted a constant with a comment.

> It would be good to beef up the comment as the byte array written by ZipEntry::setExtra is set to the unknown tag for the Tag and the length is set to 28 which includes all of the Zip64 extra fields

Opted to extract a method to create the opaque extra field since that allows for more comments and declutters the calling method.

File: test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java
Comment (eirbjo) [reply test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java]: > The length of the zip64 extra field is going to be 28. I would suggest making this a bit clearer and adding a field/constant which is set to 28 for clarity to make it easier for those who might look at this code later

I believe this should be well-described now between the constants and the `makeOpaqueExtraField` method.

**Comment by eirbjo** [comment]
A ran the latest version of the test against a reverted ZipFileSystem to confirmed that it still catches the regression it was made for.

**Comment by eirbjo** [comment]
Something worth mentioning:

This test  currently does not check that the last modified time stamp is actually not the one read from the LOC and not from the CEN. And there is a discrepancey here, `ZipFileSystem` reads it from the LOC, while ZipFile reads it from the CEN.

We could test this by setting different last modified timestamps in the LOC and CEN and use assertEquals to verify that the expected time stamp is parsed.

@LanceAndersen Do you think this is worth pursuing in this PR? I have a sketch ready, but I fear will clutter the test with too many concerns.

File: test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java
Comment (LanceAndersen) [inline test/jdk/jdk/nio/zipfs/TestLocOffsetFromZip64EF.java:191]: I might add one more entry after this one, perhaps with a compression method of DEFLATED

**Review by LanceAndersen — APPROVED** [review]
I think this looks good overall.   Please see the comment below and then I can run the test one more time and you should be good to go

**Comment by eirbjo** [comment]
This PR was approved by @LanceAndersen in March 2023, but for some reason it was never integrated. Most probably I was waiting for a final approval following the last review comments from Lance. The test was rewritten to JUnit in October 2023.

The test is useful in that it covers an aspect sensitive to the order of ZIP64 extended fields which is produced by Info-ZIP, but which ZipOutputStream cannot produce.

I think it would be good to bring this problem-listed test back to life and also make it platform independent.

**Comment by eirbjo** [comment]
This comment is just to test if nio-dev is notified of updates in this PR.

**Comment by jaikiran** [comment]
Hello Eirik, before integrating this I'll run this against a specific version of OS to verify that the test that is being removed from the problem listing does indeed work fine on that OS. Given that we no longer depend on a OS specific tool to generate the zip, I think it should work fine, but I'll do one run to be sure.

**Comment by eirbjo** [comment]
> I ran this test-only change against our CI plus specifically against Oracle Linux 9, where this test was previously failing. All tests passed. So this looks good to me.

Thanks @jaikiran. This PR was approved by Lance March last year. There have been some minor updates after that. It would be good to see an approval of this PR in its current state before I integrate.

**Review by LanceAndersen — APPROVED** [review]

**Pull request closed without merging**
